asdf/tests/_helpers.py

```diff
@@ -0,0 +1,499 @@
+
+
+def _assert_extension_type_correctness(extension, extension_type, resolver):
+    __tracebackhide__ = True
+
+    if extension_type.yaml_tag is not None and extension_type.yaml_tag.startswith(YAML_TAG_PREFIX):
+        return
+
+    if extension_type == asdf.stream.Stream:
+        # Stream is a special case.  It was implemented as a subclass of NDArrayType,
+        # but shares a tag with that class, so it isn't really a distinct type.
+        return
+
+    assert extension_type.name is not None, f"{extension_type.__name__} must set the 'name' class attribute"
+
+    # Currently ExtensionType sets a default version of 1.0.0,
+    # but we want to encourage an explicit version on the subclass.
+    assert "version" in extension_type.__dict__, "{} must set the 'version' class attribute".format(
+        extension_type.__name__,
+    )
+
+    # check the default version
+    types_to_check = [extension_type]
+
+    # Adding or updating a schema/type version might involve updating multiple
+    # packages. This can result in types without schema and schema without types
+    # for the development version of the asdf-standard. To account for this,
+    # don't include versioned siblings of types with versions that are not
+    # in one of the asdf-standard versions in supported_versions (excluding the
+    # current development version).
+    asdf_standard_versions = supported_versions.copy()
+    if asdf_standard_development_version in asdf_standard_versions:
+        asdf_standard_versions.remove(asdf_standard_development_version)
+    for sibling in extension_type.versioned_siblings:
+        tag_base, version = split_tag_version(sibling.yaml_tag)
+        for asdf_standard_version in asdf_standard_versions:
+            vm = get_version_map(asdf_standard_version)
+            if tag_base in vm["tags"] and AsdfVersion(vm["tags"][tag_base]) == version:
+                types_to_check.append(sibling)
+                break
+
+    for check_type in types_to_check:
+        schema_location = resolver(check_type.yaml_tag)
+
+        assert schema_location is not None, (
+            f"{extension_type.__name__} supports tag, {check_type.yaml_tag}, "
+            "but tag does not resolve.  Check the tag_mapping and uri_mapping "
+            f"properties on the related extension ({extension_type.__name__})."
+        )
+
+        if schema_location not in asdf.get_config().resource_manager:
+            try:
+                with generic_io.get_file(schema_location) as f:
+                    yaml.safe_load(f.read())
+            except Exception as err:  # noqa: BLE001
+                msg = (
+                    f"{extension_type.__name__} supports tag, {check_type.yaml_tag}, "
+                    f"which resolves to schema at {schema_location}, but "
+                    "schema cannot be read."
+                )
+                raise AssertionError(msg) from err
```

Comment:
Clearly downstream packages are using some of this. We cannot just hide it away like this without providing some of it back to our downstream. I thought the plan was to move this to `asdf.testing.helpers`? If so those changes need to be part of the scope of this PR.

Reply:
Offline conversation resolved this question